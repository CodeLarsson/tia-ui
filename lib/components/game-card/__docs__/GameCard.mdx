import { Canvas, Meta } from '@storybook/blocks';
import Example from './Example.tsx';
import * as GameCard from './GameCard.stories.tsx';

<Meta of={GameCard} title="GameCard" />

# GameCard

Game Card, a component that represents one of 52 cards in a playing deck

#### Example

<Canvas of={GameCard.AceOfSpades} />

## Usage

```ts
import {GameCard, CardValue, CardSuite} from "tia-ui";

const Example = () => {
  return (
      <GameCard
        value={CardValue.Ace}
        suite=(CardSuite.Spades)
        isFlipped=(false)
        isFlippbable=(true)
      />
  );
};

export default Example;
```

#### Arguments

- **value** - `CardValue` enum value
- **suite** - `CardSuite` enum value
- **isFlipped** - `bool`, if card is flipped or not
- **isFlippbable** - `bool`if card can be flipped or not
