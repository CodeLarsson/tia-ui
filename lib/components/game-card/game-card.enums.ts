export enum CardValue {
  'Two' = 2,
  'Three',
  'Four',
  'Five',
  'Six',
  'Seven',
  'Eight',
  'Nine',
  'Ten',
  'Jack',
  'Queen',
  'King',
  'Ace'
}

export enum CardSuite {
  'Hearts' = 'Hearts',
  'Diamonds' = 'Diamonds',
  'Clubs' = 'Clubs',
  'Spades' = 'Spades'
}
