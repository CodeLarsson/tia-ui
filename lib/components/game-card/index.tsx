export { default as GameCard } from './game-card';
export { CardSuite } from './game-card';
export { CardValue } from './game-card';
export type { GameCardProps } from './game-card';
