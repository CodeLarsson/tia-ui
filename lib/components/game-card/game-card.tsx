import { useState } from 'react';
import { CardSuite, CardValue } from './game-card.enums';

export type GameCardProps = {
  suite: CardSuite;
  value: CardValue;
  isFlipped?: boolean;
  isFlippable?: boolean;
  flipCard?: () => void;
};

const GameCard = ({
  suite,
  value,
  isFlipped = false,
  isFlippable = true
}: GameCardProps) => {
  const [frontUp, setFrontUp] = useState<boolean>(isFlipped);

  const getSuiteClass = (suite: CardSuite): string => {
    switch (suite) {
      case CardSuite.Hearts:
        return 'before:tia-content-card-suit-hearts tia-text-red-500';
      case CardSuite.Diamonds:
        return 'before:tia-content-card-suit-diamonds tia-text-red-500';
      case CardSuite.Clubs:
        return 'before:tia-content-card-suit-clubs tia-text-grey-900';
      case CardSuite.Spades:
        return 'before:tia-content-card-suit-spades tia-text-grey-900';
      default:
        return '';
    }
  };

  const getValueSymbol = (value: CardValue): string => {
    switch (value) {
      case CardValue.Jack:
        return 'J';
      case CardValue.Queen:
        return 'Q';
      case CardValue.King:
        return 'K';
      case CardValue.Ace:
        return 'A';
      default:
        return value.toString();
    }
  };

  const flipCard = () => {
    if (isFlippable) {
      setFrontUp(!frontUp);
    }
  };

  const suiteCssClass = getSuiteClass(suite);

  return (
    <div
      data-testid="tia-game-card"
      className={`tia-aspect-[2.5/3.5] tia-bg-white tia-w-full tia-rounded-xl md:tia-m-2 tia-flex tia-flex-col md:tia-text-2xl tia-p-1 md:tia-p-4 tia-border-black tia-shadow-md tia-transition-transform tia-duration-500 tia-transform tia-origin-center ${frontUp ? 'tia-rotate-0' : 'tia-rotate-180'} ${!isFlippable ? 'tia-pointer-events-none' : 'tia-cursor-pointer'}`}
      onClick={() => flipCard()}
    >
      <div className={`${suiteCssClass} tia-h-2 md:tia-h-12`}>
        {getValueSymbol(value)}
      </div>
      <div className="tia-flex-1 tia-flex tia-items-center">
        <div className="tia-flex tia-flex-col tia-items-center md:tia-text-4xl tia-w-full">
          <div className={suiteCssClass}></div>
        </div>
      </div>
      <div
        className={`${suiteCssClass} tia-h-2 md:tia-h-12 tia-flex tia-items-end tia-justify-end`}
      >
        {getValueSymbol(value)}
      </div>
      <div
        className={`tia-absolute tia-inset-0 tia-rounded-xl tia-h-full tia-bg-gradient-to-br tia-to-primary-500 tia-from-primary-200 tia-bg-no-repeat tia-bg-cover tia-transform-opacity tia-duration-500 ${frontUp ? 'tia-opacity-0' : 'tia-opacity-100'}`}
      ></div>
    </div>
  );
};

GameCard.displayName = 'GameCard';
export default GameCard;
