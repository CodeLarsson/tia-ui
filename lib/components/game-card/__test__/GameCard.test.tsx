import React from 'react';
import { describe, expect, it } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';

import GameCard from '../game-card';
import { CardSuite, CardValue } from '../game-card';

describe('GameCard component', () => {
  it('GameCard should render correctly', () => {
    const { container } = render(
      <GameCard suite={CardSuite.Spades} value={CardValue.Ace} />
    );
    expect(container).toMatchSnapshot();
  });

  it('GameCard should flipÃ¥ on click', () => {
    // Arrange
    render(<GameCard suite={CardSuite.Spades} value={CardValue.Ace} />);
    const card = screen.getByTestId('tia-game-card');

    // Assert Before click
    expect(card.classList).toContain('tia-rotate-180');

    // Act
    fireEvent(card, new MouseEvent('click', { bubbles: true }));

    // Assert After click
    expect(card.classList).toContain('tia-rotate-0');
  });
});
