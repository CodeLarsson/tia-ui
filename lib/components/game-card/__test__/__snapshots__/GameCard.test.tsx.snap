// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GameCard component > GameCard should render correctly 1`] = `
<div>
  <div
    class="tia-aspect-[2.5/3.5] tia-bg-white tia-w-full tia-rounded-xl md:tia-m-2 tia-flex tia-flex-col md:tia-text-2xl tia-p-1 md:tia-p-4 tia-border-black tia-shadow-md tia-transition-transform tia-duration-500 tia-transform tia-origin-center tia-rotate-180 tia-cursor-pointer"
    data-testid="tia-game-card"
  >
    <div
      class="before:tia-content-card-suit-spades tia-text-grey-900 tia-h-2 md:tia-h-12"
    >
      A
    </div>
    <div
      class="tia-flex-1 tia-flex tia-items-center"
    >
      <div
        class="tia-flex tia-flex-col tia-items-center md:tia-text-4xl tia-w-full"
      >
        <div
          class="before:tia-content-card-suit-spades tia-text-grey-900"
        />
      </div>
    </div>
    <div
      class="before:tia-content-card-suit-spades tia-text-grey-900 tia-h-2 md:tia-h-12 tia-flex tia-items-end tia-justify-end"
    >
      A
    </div>
    <div
      class="tia-absolute tia-inset-0 tia-rounded-xl tia-h-full tia-bg-gradient-to-br tia-to-primary-500 tia-from-primary-200 tia-bg-no-repeat tia-bg-cover tia-transform-opacity tia-duration-500 tia-opacity-100"
    />
  </div>
</div>
`;
